muonDIS/makeMuonDIS.py

```diff
@@ -0,0 +1,228 @@
+        kc = myPythia.Pycomp(kf)
+        myPythia.SetMDCY(kc, 1, 0)
+
+    seed = int(time.time() % 900000000)
+    myPythia.SetMRPY(1, seed)
+    mutype = {-13: "gamma/mu+", 13: "gamma/mu-"}
+
+    myPythia.SetMSTU(11, 11)
+    logging.info(f"Processing events from {nStart} to {nEnd-1}...")
+
+    nMade = 0
+
+    for k in range(nStart, nEnd):
+        muon_tree.GetEvent(k)
+
+        imuondata = muon_tree.imuondata
+
+        pid = imuondata[0]
+        px = imuondata[1]
+        py = imuondata[2]
+        pz = imuondata[3]
+        x = imuondata[4]
+        y = imuondata[5]
+        z = imuondata[6]
+        w = imuondata[7]
+        time_muon = imuondata[8]
+
+        p = r.TMath.Sqrt(px**2 + py**2 + pz**2)
+        mass = PDG.GetParticle(abs(int(pid))).Mass()
+        E = r.TMath.Sqrt(mass**2 + p**2)
+
+        theta = r.TMath.ACos(pz / p)
+        phi = r.TMath.ATan2(py, px)
+
+        logging.debug(
+            f"\n\tMuon index:{k} \n\tPID = {pid}, weight = {w} \n\tpx = {px}, py = {py}, pz = {pz}, E = {E},\n\tx = {x}, y = {y}, z = {z}\n"
+        )
+
+        isProton = 1
+        xsec = 0
+
+        mu = array("d", [pid, px, py, pz, E, x, y, z, w, isProton, xsec, time_muon])
+        muPart = r.TVectorD(12, mu)
+        myPythia.Initialize("FIXT", mutype[pid], "p+", p)  # target = "p+"
+        myPythia.Pylist(1)
+
+        for a in range(args.nDIS):
+            if a == args.nDIS // 2:
+                myPythia.Initialize("FIXT", mutype[pid], "n0", p)  # target = "n0"
+                isProton = 0
+                logging.debug("Switching to neutron interaction")
+
+            dPartDIS.Clear()
+            iMuon.Clear()
+            muPart[9] = isProton
+            iMuon[0] = muPart
+            myPythia.GenerateEvent()
+            myPythia.Pyedit(1)
+            logging.debug(
+                f"DIS Event {a} generated, number of particles: {myPythia.GetN()}"
+            )
+
+            for itrk in range(1, myPythia.GetN() + 1):
+                xsec = myPythia.GetPARI(1)
+                muPart[10] = xsec
+                did = myPythia.GetK(itrk, 2)
+                dpx, dpy, dpz = rotate(
+                    myPythia.GetP(itrk, 1),
+                    myPythia.GetP(itrk, 2),
+                    myPythia.GetP(itrk, 3),
+                    theta,
+                    phi,
+                )
+                psq = dpx**2 + dpy**2 + dpz**2
+                masssq = PDG.GetParticle(did).Mass() ** 2
+                E = r.TMath.Sqrt(masssq + psq)
+                m = array("d", [did, dpx, dpy, dpz, E])
+                part = r.TVectorD(5, m)
+                nPart = dPartDIS.GetEntries()
+                if dPartDIS.GetSize() == nPart:
+                    dPartDIS.Expand(nPart + 10)
+                # dPartDIS.ConstructedAt(nPart).Use(part) #to be adapted later
+                dPartDIS[nPart] = part
+                if itrk == 1:
+                    with open(f"sigmadata_{args.nJob}.txt", "a") as fcross:
```

Comment:
Reasoning for having an external textfile to save the DIS cross section from Pythia6:
1) As is, the DIS cross section otherwise is not saved anywhere
2) Consider generating N >>1 DIS events for the single muon. Since Pythia generates the cross-section value events-by-event, it converges from some meaningless value for the 1st event to the proper value for the Nth event. In the proper implementation, all the events must share the same correct cross-section. As a temporary fix,we achieve this by reading the Nth DIS cross section entry for a given muon to calculate the weight.

Reply:
My plan of action to avoid this external file is:

1.Directly write the cross section to the muonDis file ( which is passed as the input for muonDIS production)
2. Set weight of the scattered muon to the cross section in the simulation (atm both the incoming and scattered have the same weight from the spill).

Perhaps also set a minimum value for the number of DIS per muon to allow for convergence?

Please let me know what you think! @olantwin

Reply:
Heres a sample from 200 DIS events per muon.
![Crosssections](https://github.com/user-attachments/assets/2fb0d547-9b15-4600-a2cd-55daeefd5564)

Reply:
Ok, this is now much clearer, thanks!

I agree, that this information should be directly in the sample. Is there a weight field in the event header? Otherwise, using the weight of the scattered muon seems like an OK workaround (maybe we can preserve the weight of the incoming muon from the target sim).

We do not use the cross section anywhere else, right?

Could you prepare that plot also relative to the final cross section? That allow us to estimate a good default number of events to generate.

Reply:
> I agree, that this information should be directly in the sample. Is there a weight field in the event header? 
I couldnt find one in ShipEventHeader.
> Otherwise, using the weight of the scattered muon seems like an OK workaround (maybe we can preserve the weight of the incoming muon from the target sim).
At the moment I save the cross section onto the scattered muon and also refactor the incoming muon's weight to include the DIS multiplicity such that we maintain the event weight's normalisation to a spill. Previously this refactoring was explicitly done in the analysis, but the meta data (nDIS per muon) needed to be explicitly plugged in to calculate weight for any sim file.

> We do not use the cross section anywhere else, right?
> 
> Could you prepare that plot also relative to the final cross section? That allow us to estimate a good default number of events to generate.